packages/php-wasm/compile/php/php_wasm.c

```diff
@@ -102,17 +102,43 @@ EMSCRIPTEN_KEEPALIVE FILE *wasm_popen(const char *cmd, const char *mode)
 			return 0;
 		}
 
+        int **descv = malloc(sizeof(int *) * 3);
+
+        int *stdin = malloc(sizeof(int) * 3);
+        int *stdout = malloc(sizeof(int) * 3);
+        int *stderr = malloc(sizeof(int) * 3);
+
+        stdin[0] = 0;
+        stdin[1] = stdin_childend;
+        stdin[2] = (int) NULL;
+
+        stdout[0] = 1;
+        stdout[1] = stdout_pipe[0];
+        stdout[2] = stdout_pipe[1];
+
+        stderr[0] = 2;
+        stderr[1] = stderr_pipe[0];
+        stderr[2] = stderr_pipe[1];
+
+        descv[0] = stdin;
+        descv[1] = stdout;
+        descv[2] = stderr;
```

Comment:
This is really cool and readable!

Reply:
You're right, I will modify this.

Reply:
@adamziel In fact, in the actual code, the position of the `descv` `malloc` allocation is important. Moving the variable made the test `popen("cat", "w")` in PHP `8.1` fail because first array pointer magically moved from 98... to 40... . I need to keep these elements allocation in that order.